src/goto-instrument/code_contracts.cpp

```diff
@@ -370,25 +402,74 @@ void code_contractst::add_contract_check(
   dest.destructive_insert(dest.instructions.begin(), check);
 }
 
-void code_contractst::operator()()
+bool code_contractst::replace_calls(
+  const std::list<std::string> &funs_to_replace)
 {
-  Forall_goto_functions(it, goto_functions)
-    code_contracts(it->second);
-
-  goto_functionst::function_mapt::iterator i_it=
-    goto_functions.function_map.find(INITIALIZE_FUNCTION);
-  assert(i_it!=goto_functions.function_map.end());
+  bool fail = false;
+  Forall_goto_functions(fit, goto_functions)
+  {
+    Forall_goto_program_instructions(it, fit->second.body)
+    {
+      if(it->is_function_call())
+      {
+        const code_function_callt &call = it->get_function_call();
+
+        // TODO we don't handle function pointers
+        if(call.function().id() != ID_symbol)
+          continue;
+
+        const irep_idt &fun_name =
+          to_symbol_expr(call.function()).get_identifier();
+        auto found = std::find(
+          funs_to_replace.begin(), funs_to_replace.end(), id2string(fun_name));
+        if(found == funs_to_replace.end())
+          continue;
+
+        if(!has_contract(fun_name))
```

Comment:
Given that the only way to get to this block of code is if fun_name is in funs_to_replace, it would be better from a performance point of view to move this out of the nest loops and just do an initial iteration through funs_to_replace before even attempting to do the callsite scanning. Of course, that leads to slightly different user behaviour... If this is done outside the loops they get the error message for any function in funs_to_replace with no contract (regardless of whether they get called anywhere or not), whereas if it's done here inside the loops, they only get the error for any of those that are actually _called_.